scripts/codegen/codegen.py

```diff
@@ -0,0 +1,611 @@
+            if lines[0].strip() == '':
+                del lines[0]
+            if lines and lines[-1].strip() == '':
+                del lines[-1]
+            output = '\n'.join(lines)+'\n'
+        if dedent:
+            output = reindentBlock(output)
+        self._outstring += output
+
+    ##
+    # @brief Extract the executable Python code from the generator.
+    #
+    def _get_code(self, fname, snippet_offset):
+        # If the markers and lines all have the same prefix
+        # (end-of-line comment chars, for example),
+        # then remove it from all the lines.
+        common_prefix = os.path.commonprefix(self.markers + self.lines)
+        if not common_prefix:
+            # there may be a prefix error
+            # just do some heuristics
+            if fname.endswith(
+                ('.h', '.hxx', '.c', '.cpp', '.cxx')):
+                # assume C/C++ comments
+                for line in (self.markers + self.lines):
+                    if line.strip().startswith('*'):
+                        common_prefix = '*'
+                    elif line.strip().startswith('//'):
+                        common_prefix = '//'
+                        break
+            if common_prefix:
+                # there should be a common prefix -> error
+                lines = list() # lines with correct prefix
+                for lineno, line in enumerate(self.lines):
+                    if not line.strip().startswith(common_prefix):
+                        print("Codegen: Comment prefix may miss in codegen snippet (+{}) in '{}'.".format(
+                            snippet_offset, fname))
+                        line_start = lineno - 5
+                        if line_start < 0:
+                            line_start = 0
+                        line_end = lineno + 5
+                        if line_end > len(self.lines):
+                            line_end = len(self.lines)
+                        for i in range(line_start, line_end):
+                            snippet_lineno = i + 2
+                            input_lineno = snippet_offset + snippet_lineno
+                            print("#{} (+{}, line {}): {}".format(
+                                input_lineno, snippet_offset, snippet_lineno, self.lines[i]))
+                    else:
+                        lines.append(line)
+                if len(lines) >= int(len(self.lines) / 2):
+                    common_prefix = os.path.commonprefix(lines)
+                    print("Codegen: Assuming comment prefix '{}' for codegen snippet (+{}) in '{}'.".format(
+                          common_prefix, snippet_offset, fname))
+                else:
+                    common_prefix = ''
+        if common_prefix:
+            self.markers = [ line.replace(common_prefix, '', 1) for line in self.markers ]
+            self.lines = [ line.replace(common_prefix, '', 1) for line in self.lines ]
+
+        return reindentBlock(self.lines, '')
```

Comment:
Do error checking for lines that have a tab in the whitespace at the beginning of a line. Try to replace tab by 8 character aligned spaces taking the length of the already removed common prefix into account. Generate a warning message.